File: Python/tigre/utilities/filtering.py
from __future__ import division
from __future__ import print_function
from numpy.core.arrayprint import dtype_is_implied
from tigre.utilities.parkerweight import parkerweight
import numpy as np
from scipy.fft  import fft, ifft

import warnings

import numpy as np
from tigre.utilities.parkerweight import parkerweight


# TODO: Fix parker
def filtering(proj, geo, angles, parker, verbose=False):
    if parker:
        proj=parkerweight(proj.transpose(0,2,1),geo,angles,parker).transpose(0,2,1)

    filt_len=max(64,2**nextpow2(2*max(geo.nDetector)))
    ramp_kernel=ramp_flat(filt_len)

    d=1
    filt=filter(geo.filter,ramp_kernel[0],filt_len,d,verbose=verbose)
    filt=np.kron(np.ones((np.int64(geo.nDetector[0]),1),dtype=np.float32),filt)

    padding = int((filt_len-geo.nDetector[1])//2 )
    scale_factor = (geo.DSD[0]/geo.DSO[0]) * (2 * np.pi/ len(angles)) / ( 4 * geo.dDetector[0] ) 

    #filter 2 projection at a time packing in to complex container
    fproj=np.empty((geo.nDetector[0],filt_len),dtype=np.complex64)
    for i in range(0,angles.shape[0]-1,2):
        fproj.fill(0)
        fproj.real[:,padding:padding+geo.nDetector[1]]=proj[i]
        fproj.imag[:,padding:padding+geo.nDetector[1]]=proj[i+1]

        fproj=fft(fproj,axis=1)
        fproj=fproj*filt
        fproj=ifft(fproj,axis=1)

        proj[i]=fproj.real[:,padding:padding+geo.nDetector[1]] * scale_factor
        proj[i+1]=fproj.imag[:,padding:padding+geo.nDetector[1]] * scale_factor

    #if odd number of projections filter last solo
    if angles.shape[0] % 2:
        fproj.fill(0)
        fproj.real[:,padding:padding+geo.nDetector[1]]=proj[angles.shape[0]-1]

        fproj=fft(fproj,axis=1)
        fproj=fproj*filt
        fproj=np.real(ifft(fproj,axis=1))     
        proj[angles.shape[0]-1]=fproj[:,padding:padding+geo.nDetector[1]] * scale_factor

    return proj


def ramp_flat(n, verbose=False):
    nn = np.arange(-n / 2, n / 2)
    h = np.zeros(nn.shape, dtype=np.float32)
    h[int(n / 2)] = 1 / 4
    odd = nn % 2 == 1
    h[odd] = -1 / (np.pi * nn[odd]) ** 2
    return h, nn


def filter(filter, kernel, order, d, verbose=False):
    f_kernel = abs(np.fft.fft(kernel)) * 2

    filt = f_kernel[: int((order / 2) + 1)]
    w = 2 * np.pi * np.arange(len(filt)) / order

    if filter in {"ram_lak", None}:
        if filter is None and verbose:
            warnings.warn("no filter selected, using default ram_lak")
    elif filter == "shepp_logan":
        filt[1:] *= np.sin(w[1:] / (2 * d)) / (w[1:] / (2 * d))
    elif filter == "cosine":
        filt[1:] *= np.cos(w[1:] / (2 * d))
    elif filter == "hamming":
        filt[1:] *= 0.54 + 0.46 * np.cos(w[1:] / d)
    elif filter == "hann":
        filt[1:] *= (1 + np.cos(w[1:]) / d) / 2
    else:
        raise ValueError("filter not recognised: " + str(filter))

    filt[w > np.pi * d] = 0
    filt = np.hstack((filt, filt[1:-1][::-1]))
    return filt.astype(np.float32)


def nextpow2(n):
    i = 1
    while (2 ** i) < n:
        i += 1
    return i

File: MATLAB/Utilities/filtering.m
function [ proj ] = filtering(proj,geo,angles,parker,varargin)
%FILTERING Summary of this function goes here
%   Detailed explanation goes here
%--------------------------------------------------------------------------
%--------------------------------------------------------------------------
% This file is part of the TIGRE Toolbox
% 
% Copyright (c) 2015, University of Bath and 
%                     CERN-European Organization for Nuclear Research
%                     All rights reserved.
%
% License:            Open Source under BSD. 
%                     See the full license at
%                     https://github.com/CERN/TIGRE/blob/master/LICENSE
%                     and
%                     https://www.mathworks.com/matlabcentral/fileexchange/view_license?file_info_id=35548
%
% Contact:            tigre.toolbox@gmail.com
% Codes:              https://github.com/CERN/TIGRE/
% Coded by:           Kyungsang Kim, modified by Ander Biguri 
%--------------------------------------------------------------------------
[usegpufft,gpuids]=parse_inputs(varargin);

if parker
	proj = ParkerWeight(proj,geo,angles,parker);
	diff_angles = diff(angles(1,:)); angle_step = mean(abs(diff_angles)); % to be used later
end 

filt_len = max(64,2^nextpow2(2*geo.nDetector(1)));
[ramp_kernel] = ramp_flat(filt_len);

d = 1; % cut off (0~1)
[filt] = Filter(geo.filter, ramp_kernel, filt_len, d);
if usegpufft>0
    filt = filt';
else
    filt = repmat(filt',[1 geo.nDetector(2)]);
end

proj=permute(proj,[2 1 3]);

if usegpufft==2
    bundle_size = 32;  %len(gpuids)
    n_bundles = floor((size(angles,2)+bundle_size-1)/bundle_size);
    n_angles = size(angles, 2);
    for ii=1:n_bundles
        
        if ii ~= n_bundles
            bundle_size_actual = bundle_size;
        else
            bundle_size_actual = n_angles-(ii-1)*bundle_size;
        end
        idx_begin = (ii-1)*bundle_size+1;
        idx_end = (ii-1)*bundle_size+bundle_size_actual+1;
        proj_flt = ApplyPaddingAndFbpFiltration(proj, idx_begin, idx_end, filt, 1, gpuids.devices);
        bundle_range = idx_begin:(idx_end-1);
        if parker
            proj_flt = proj_flt/2/geo.dDetector(1)*(2*pi/  (pi/angle_step)  )/2*(geo.DSD(bundle_range)/geo.DSO(bundle_range));
        else
            proj_flt = proj_flt/2/geo.dDetector(1)*(2*pi/  size(angles,2)   )/2*(geo.DSD(bundle_range)/geo.DSO(bundle_range));
        end
        proj(:,:,bundle_range) = proj_flt;
    end
else 
    
    for ii=1:size(angles,2)

        fproj = (zeros(filt_len,geo.nDetector(2),'single'));

        fproj(round(filt_len/2-geo.nDetector(1)/2+1):round(filt_len/2+geo.nDetector(1)/2),:) = proj(:,:,ii);

        fproj = fft(fproj);   
        fproj = fproj.*filt;
        fproj = (real(ifft(fproj)));

        if parker
            proj(:,:,ii) = fproj(round(end/2-geo.nDetector(1)/2+1):round(end/2+geo.nDetector(1)/2),:)/2/geo.dDetector(1)*(2*pi/  (pi/angle_step)   )/2*(geo.DSD(ii)/geo.DSO(ii));
        else
            proj(:,:,ii) = fproj(round(end/2-geo.nDetector(1)/2+1):round(end/2+geo.nDetector(1)/2),:)/2/geo.dDetector(1)*(2*pi/  size(angles,2)   )/2*(geo.DSD(ii)/geo.DSO(ii));
        end 

    end
end

proj=permute(proj,[2 1 3]);
end

function [h, nn] = ramp_flat(n)
nn = (-(n/2):(n/2-1))';
h = zeros(size(nn),'single');
h(n/2+1) = 1 / 4;
odd = mod(nn,2) == 1;
h(odd) = -1 ./ (pi * nn(odd)).^2;
end


function [filt] = Filter(filter, kernel, order, d)

f_kernel = abs(fft(kernel))*2;
filt = f_kernel(1:order/2+1)';
w = 2*pi*(0:size(filt,2)-1)/order;   % frequency axis up to Nyquist 

switch lower(filter)
    case 'ram-lak'
        % Do nothing
    case 'shepp-logan'
        % be careful not to divide by 0:
        filt(2:end) = filt(2:end) .* (sin(w(2:end)/(2*d))./(w(2:end)/(2*d)));
    case 'cosine'
        filt(2:end) = filt(2:end) .* cos(w(2:end)/(2*d));
    case 'hamming'  
        filt(2:end) = filt(2:end) .* (.54 + .46 * cos(w(2:end)/d));
    case 'hann'
        filt(2:end) = filt(2:end) .*(1+cos(w(2:end)./d)) / 2;
    otherwise
        disp(filter);
        error('Invalid filter selected.');
end

filt(w>pi*d) = 0;                      % Crop the frequency response
filt = [filt , filt(end-1:-1:2)];    % Symmetry of the filter
return

end

function [usegpufft, gpuids]=parse_inputs(argin)

opts =  {'usegpufft','gpuids'};
defaults=ones(length(opts),1);

% Check inputs
nVarargs = length(argin);
if mod(nVarargs,2)
    error('TIGRE:filtering:InvalidInput','Invalid number of inputs')
end

% check if option has been passed as input
for ii=1:2:nVarargs
    ind=find(ismember(opts,lower(argin{ii})));
    if ~isempty(ind)
        defaults(ind)=0;
    end
end

for ii=1:length(opts)
    opt=opts{ii};
    default=defaults(ii);
    % if one option is not default, then extract value from input
    if default==0
        ind=double.empty(0,1);jj=1;
        while isempty(ind)
            ind=find(isequal(opt,lower(argin{jj})));
            jj=jj+1;
        end
        if isempty(ind)
            error('TIGRE:filtering:InvalidInput',['Optional parameter "' argin{jj} '" does not exist' ]); 
        end
        val=argin{jj};
    end
    
    switch opt
        case 'usegpufft'
            if default
                usegpufft=2;
            else
                usegpufft=val;
            end            
        case 'gpuids'
            if default
                gpuids = GpuIds();
            else
                gpuids = val;
            end
       
        otherwise
            error('TIGRE:filtering:InvalidInput',['Invalid input name:', num2str(opt),'\n No such option in FDK()']);
    end
end

end
